src/query/ts/m3db/consolidators/step_accumulator.go

```diff
@@ -61,35 +56,28 @@ func NewStepLookbackAccumulator(
 	}
 }
 
-// AddPointForIterator adds a datapoint to a given step if it's within the valid
+// AddPoint adds a datapoint to a given step if it's within the valid
 // time period; otherwise drops it silently, which is fine for accumulation.
-func (c *StepLookbackAccumulator) AddPointForIterator(
-	dp ts.Datapoint,
-	i int,
-) {
+func (c *StepLookbackAccumulator) AddPoint(dp ts.Datapoint) {
 	if dp.Timestamp.Before(c.earliestLookback) {
 		// this datapoint is too far in the past, it can be dropped.
 		return
 	}
 
-	c.datapoints[i] = append(c.datapoints[i], xts.Datapoint{
+	c.datapoints = append(c.datapoints, xts.Datapoint{
 		Timestamp: dp.Timestamp,
 		Value:     dp.Value,
 	})
 }
 
 // AccumulateAndMoveToNext consolidates the current values and moves the
 // consolidator to the next given value, purging stale values.
-func (c *StepLookbackAccumulator) AccumulateAndMoveToNext() []xts.Datapoints {
+func (c *StepLookbackAccumulator) AccumulateAndMoveToNext() []xts.Datapoint {
 	// Update earliest lookback then remove stale values for the next
 	// iteration of the datapoint set.
 	c.earliestLookback = c.earliestLookback.Add(c.stepSize)
-	accumulated := make([]xts.Datapoints, len(c.datapoints))
-	for i, dps := range c.datapoints {
-		accumulated[i] = make(xts.Datapoints, len(dps))
-		copy(accumulated[i], dps)
-		c.datapoints[i] = c.datapoints[i][:0]
-	}
-
+	accumulated := make([]xts.Datapoint, len(c.datapoints))
```

Comment:
Is it plausible to pool these allocations at some point in the future?  

I'm still trying to understand the lifecycle better here, so I'm definitely out of my depth in being able to say one way or another whether it's a good idea hah.

Reply:
Slightly simplified flow for this is that the `N-th` iterator in the encoded step iterator has it's `Next()` called, adding those points to the `N-th StepCollector`, until it either

a) returns false, in which case the accumulated values are consolidated by the `StepCollector` and set to the `Current` values

b) yields a point whose timestamp is beyond the current step, in which case that point is saved as a "peeked value" in preparation for the next time `Next()` is called, and the accumulated values are consolidated as for the above case.

We definitely should add pooling here but it may be a fairly involved change largely because we don't have a great lifecycle management story for them yet

Reply:
In terms of the different `StepCollectors`, the step iterator variant that uses `StepLookbackAccumulator` is the 
"unconsolidated" one, which just gives back all the raw points. These are used only in temporal functions like `sum_over_time`, since they need to work on the rawest data available. All othe functions use the `StepLookbackConolidator`, which gives "consolidated" data points, which are snapped to stepSize and have a given consolidation function applied to them.

Frankly, I really dislike the concept of having this consolidated/unconsolidated split as is; I think we may be able to add a tiny bit of  carefully placed logic to slot in a specialized consolidation function where necessary (also that will  also give us `summarize` for free).